heron/schedulers/src/java/com/twitter/heron/scheduler/mesos/MesosScheduler.java

```diff
@@ -0,0 +1,237 @@
+      // i.e. waits for the message sent to Mesos Master
+      long stopTimeoutInMs = MesosContext.getHeronMesosSchedulerDriverStopTimeoutMs(config);
+      joinSchedulerDriver(stopTimeoutInMs, TimeUnit.MILLISECONDS);
+    }
+
+    // Will not kill the topology when close() is invoked,
+    // since the lifecycle of Topology is independent from Scheduler
+  }
+
+  @Override
+  public boolean onSchedule(PackingPlan packing) {
+    // Construct the jobDefinition
+    Map<Integer, BaseContainer> jobDefinition = new HashMap<>();
+
+    for (int containerIndex = 0;
+         containerIndex < Runtime.numContainers(runtime);
+         containerIndex++) {
+      jobDefinition.put(containerIndex, getBaseContainer(containerIndex, packing));
+    }
+
+    return mesosFramework.createJob(jobDefinition);
+  }
+
+  @Override
+  public List<String> getJobLinks() {
+    Protos.FrameworkID frameworkID = mesosFramework.getFrameworkId();
+    // FrameworkID should exist otherwise onSchedule(..) returned false directly earlier
+    // So no need to null check
+
+    // The job link's format
+    String jobLink = String.format("%s/#/frameworks/%s",
+        MesosContext.getHeronMesosMasterUri(config), frameworkID.getValue());
+
+    List<String> jobLinks = new ArrayList<>();
+    jobLinks.add(jobLink);
+    return jobLinks;
+  }
+
+  @Override
+  public boolean onKill(Scheduler.KillTopologyRequest request) {
+    return mesosFramework.killJob();
+  }
+
+  @Override
+  public boolean onRestart(Scheduler.RestartTopologyRequest request) {
+    int containerId = request.getContainerIndex();
+    return mesosFramework.restartJob(containerId);
+  }
+
+  protected MesosFramework getMesosFramework() {
+    return new MesosFramework(config, runtime);
+  }
+
+  protected SchedulerDriver getSchedulerDriver(String masterURI, MesosFramework framework) {
+    Protos.FrameworkInfo.Builder frameworkBuilder = Protos.FrameworkInfo.newBuilder()
+        .setUser("") // Have Mesos fill in the current user.
+        .setName("heron_scheduler_" + Context.topologyName(config))
+        .setCheckpoint(true);
+
+    return new MesosSchedulerDriver(
+        framework,
+        frameworkBuilder.build(),
+        masterURI);
+  }
+
+
+  /**
+   * Get BaseContainer info.
+   *
+   * @param containerIndex container index to start
+   * @param packing the PackingPlan
+   * @return BaseContainer Info
+   */
+  protected BaseContainer getBaseContainer(Integer containerIndex, PackingPlan packing) {
+    BaseContainer container = new BaseContainer();
+
+    container.name = TaskUtils.getTaskNameForContainerIndex(containerIndex);
+
+    container.runAsUser = Context.role(config);
+    container.description = String.format("Container %d for topology %s",
+        containerIndex, Context.topologyName(config));
+
+    // We assume every container is homogeneous
+    PackingPlan.Resource containerResource =
+        packing.containers.values().iterator().next().resource;
```

Comment:
Instead of assuming that all packing plans will always return homogeneous containers, this method could find max container size (one time computation) and use it?